src/invariant/InvariantManagerImpl.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "invariant/InvariantDoesNotHold.h"
+#include "invariant/InvariantManagerImpl.h"
+#include "invariant/CacheIsConsistentWithDatabase.h"
+#include "invariant/ChangedAccountsSubentriesCountIsValid.h"
+#include "invariant/TotalCoinsEqualsBalancesPlusFeePool.h"
+#include "ledger/LedgerDelta.h"
+#include "lib/util/format.h"
+#include "util/Logging.h"
+#include "xdrpp/printer.h"
+#include <memory>
+#include <numeric>
+
+namespace stellar
+{
+
+std::unique_ptr<InvariantManager>
+InvariantManager::create(Application& app)
+{
+    return make_unique<InvariantManagerImpl>();
+}
+
+InvariantManagerImpl::InvariantManagerImpl()
+{
+}
+
+void
+InvariantManagerImpl::checkOnLedgerClose(
+        TxSetFramePtr const& txSet,
+        LedgerDelta const& delta)
+{
+    for (auto invariant : mEnabled)
+    {
+        auto result = invariant->checkOnLedgerClose(delta);
+        if (result.empty())
+        {
+            continue;
+        }
+
+        auto transactions = TransactionSet{};
+        txSet->toXDR(transactions);
+        auto message =
+            fmt::format(R"(invariant "{}" does not hold on ledger {}: {}{}{})",
+                        invariant->getName(), delta.getHeader().ledgerSeq, result,
+                        "\n", xdr::xdr_to_string(transactions));
+        CLOG(FATAL, "Invariant") << message;
```

Comment:
note that CLOG followed by throw is typically an anti-pattern as we expect the caller to log (in this case, if it does we end up with multiple log lines with the same stuff).
I think this one might be OK as it will crash the app, the others are not as there is already logic to catch exceptions during initialization.